### PR TITLE
[WIP] Migrations

### PR DESCRIPTION
**Pull request by mikeminutillo**
- Add the concept of migrations
- Add a migration to detect problematic failed messages and split them up

### CODE REVIEW

**Comment by mikeminutillo** [comment]
@johnsimons thoughts?

I need to write some tests for these

File: src/ServiceControl/DbMigrations/1.27/SplitFailedMessageDocumentsMigration.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                {
+                    Check(failedMessage, session);
+                }
+
+            } while (retrievedResults == PageSize);
+        }
+
+        private void Check(FailedMessage failedMessage, IDocumentSession session)
+        {
+            var originalRecords = failedMessage.ProcessingAttempts.Select(x => new ProcessingAttemptRecord(x)).ToArray();
```
Comment (johnsimons) [inline src/ServiceControl/DbMigrations/1.27/SplitFailedMessageDocumentsMigration.cs]: don't we need to filer based on `Unresolved` only ?

File: src/ServiceControl/DbMigrations/1.27/SplitFailedMessageDocumentsMigration.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                {
+                    // We've retried to this endpoint so it's the primary
+                    continue;
+                }
+
+                var newFailedMessage = new FailedMessage
+                {
+                    Id = FailedMessage.MakeDocumentId(group.Key),
+                    UniqueMessageId = group.Key,
+                    ProcessingAttempts = group.Select(x => x.Attempt).OrderBy(x => x.FailureDetails.TimeOfFailure).ToList(),
```
Comment (johnsimons) [inline src/ServiceControl/DbMigrations/1.27/SplitFailedMessageDocumentsMigration.cs]: I don't get this? When would there be more then one attempt ?

File: src/ServiceControl/DbMigrations/1.27/SplitFailedMessageDocumentsMigration.cs (hunk 1 above)
Comment (mikeminutillo) [inline src/ServiceControl/DbMigrations/1.27/SplitFailedMessageDocumentsMigration.cs]: What if we have a FailedMessage that we retried but it has one of these duplicates in it? Shouldn't we pull the duplicate out?

File: src/ServiceControl/DbMigrations/1.27/SplitFailedMessageDocumentsMigration.cs (hunk 2 above)
Comment (mikeminutillo) [inline src/ServiceControl/DbMigrations/1.27/SplitFailedMessageDocumentsMigration.cs]: When we've performed at least one retry there will be multiple processing attempts

**Pull request merged**